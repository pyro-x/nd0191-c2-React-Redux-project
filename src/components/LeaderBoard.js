import { connect } from 'react-redux';

const LeaderBoard = ({usersWithScore}) => {
    return (
        <div className='border' > 
            <div className="grid grid-flow-row grid-cols-3 gap-3 bg-blue-200 border divide-x">
                <div className="p-2">Name</div>
                <div className="p-2">Answered</div>
                <div className="p-2">Created</div>
            </div>
                {
                    usersWithScore.sort((a,b) => b.score - a.score).map((user) => {
                        return (
                                <div key={user.id} data-testid="user-row" className="grid grid-flow-row grid-cols-3 gap-4 boder divide-x">
                                    <div className="p-2 flex items-center"><img src={user.avatarURL} alt="avatar" className="w-10 h-10 rounded-full pr-2" />{user.name}</div>
                                    <div className="p-2 flex items-center" data-testid="nquestion-answered">{user.answeredQuestions}</div>
                                    <div className="p-2 flex items-center" data-testid="nquestion-created">{user.createdQuestions}</div>
                                </div>
                        )
                    })
                }    
        </div>
    )
}

const mapStateToProps = ({ users }) => {
    const  usersWithScore = Object.values(users).map((user) => {

        const answeredQuestions = Object.keys(user.answers).length;
        const createdQuestions = user.questions.length;
        const score = answeredQuestions + createdQuestions;
        // return as an object with the user id as the key
        
        return {...user, answeredQuestions, createdQuestions, score} ;
        
    });

    return {
        usersWithScore
    }
};



export default connect(mapStateToProps)(LeaderBoard);
