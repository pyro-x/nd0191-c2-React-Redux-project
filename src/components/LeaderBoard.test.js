
import * as React from 'react';
import {  render, screen } from '@testing-library/react';

import reducer from "../reducers";
import { Provider } from 'react-redux';
import middleware from "../middleware";
import { createStore } from 'redux';
import { MemoryRouter } from 'react-router';

import LeaderBoard from "./LeaderBoard";

describe ('LeaderBoard', () => {
    let store;
    const initialState = {
        authedUser: "sarahedo",
        users: {
            sarahedo: {
                id: "sarahedo",
                name: "Sarah Edo",
                avatarURL: "https://api.dicebear.com/5.x/avataaars/svg?seed=sarahedo&style=circle",
                answers: {
                    "8xf0y6ziyjabvozdd253nd": "optionOne",
                    "6ni6ok3ym7mf1p33lnez": "optionOne",
                    "am8ehyc8byjqgar0jgpub9": "optionTwo",
                    "loxhs1bqm25b708cmbf3g": "optionTwo"
                },
                questions : ['8xf0y6ziyjabvozdd253nd', 'am8ehyc8byjqgar0jgpub9'],
                password: "password123"
            },
            tylermcginnis: {
                id: "tylermcginnis",
                name: "Tyler McGinnis",
                avatarURL: "https://api.dicebear.com/5.x/avataaars/svg?seed=tylermcginnis&style=circle",
                answers: {
                    "vthrdm985a262al8qx3do": "optionOne",
                    "xj352vofupe1dqz9emx13r": "optionTwo",
                },
                questions : ['loxhs1bqm25b708cmbf3g', 'vthrdm985a262al8qx3do'],
                password: "password123"
            }
        }
    }

    beforeEach(() => {
        store = createStore(reducer, initialState, middleware);
    });

    it ('renders correctly', () => {
        const  view  = render(
            <MemoryRouter>
                <Provider store={store}>
                    <LeaderBoard/>
                </Provider>
            </MemoryRouter>
            );
        expect(view).toMatchSnapshot()
    });


    it('render two users', () => {
        const view = render(
            <MemoryRouter>
                <Provider store={store}>
                    <LeaderBoard/>
                </Provider>
            </MemoryRouter>
        );

        expect(screen.getAllByTestId('user-row')).toHaveLength(2);

        expect(screen.getAllByTestId('user-row')[0]).toHaveTextContent('Sarah Edo');
        expect(screen.getAllByTestId('user-row')[1]).toHaveTextContent('Tyler McGinnis');

        expect(screen.getAllByTestId('nquestion-answered')[0]).toHaveTextContent('4');
        expect(screen.getAllByTestId('nquestion-created')[0]).toHaveTextContent('2');


    });
});
    