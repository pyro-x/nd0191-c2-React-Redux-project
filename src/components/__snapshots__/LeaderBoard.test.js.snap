// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaderBoard renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="border"
      >
        <div
          class="grid grid-flow-row grid-cols-3 gap-3 bg-blue-200 border divide-x"
        >
          <div
            class="p-2"
          >
            Name
          </div>
          <div
            class="p-2"
          >
            Answered
          </div>
          <div
            class="p-2"
          >
            Created
          </div>
        </div>
        <div
          class="grid grid-flow-row grid-cols-3 gap-4 boder divide-x"
          data-testid="user-row"
        >
          <div
            class="p-2 flex items-center"
          >
            <img
              alt="avatar"
              class="w-10 h-10 rounded-full pr-2"
              src="https://api.dicebear.com/5.x/avataaars/svg?seed=sarahedo&style=circle"
            />
            Sarah Edo
          </div>
          <div
            class="p-2 flex items-center"
            data-testid="nquestion-answered"
          >
            4
          </div>
          <div
            class="p-2 flex items-center"
            data-testid="nquestion-created"
          >
            2
          </div>
        </div>
        <div
          class="grid grid-flow-row grid-cols-3 gap-4 boder divide-x"
          data-testid="user-row"
        >
          <div
            class="p-2 flex items-center"
          >
            <img
              alt="avatar"
              class="w-10 h-10 rounded-full pr-2"
              src="https://api.dicebear.com/5.x/avataaars/svg?seed=tylermcginnis&style=circle"
            />
            Tyler McGinnis
          </div>
          <div
            class="p-2 flex items-center"
            data-testid="nquestion-answered"
          >
            2
          </div>
          <div
            class="p-2 flex items-center"
            data-testid="nquestion-created"
          >
            2
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="border"
    >
      <div
        class="grid grid-flow-row grid-cols-3 gap-3 bg-blue-200 border divide-x"
      >
        <div
          class="p-2"
        >
          Name
        </div>
        <div
          class="p-2"
        >
          Answered
        </div>
        <div
          class="p-2"
        >
          Created
        </div>
      </div>
      <div
        class="grid grid-flow-row grid-cols-3 gap-4 boder divide-x"
        data-testid="user-row"
      >
        <div
          class="p-2 flex items-center"
        >
          <img
            alt="avatar"
            class="w-10 h-10 rounded-full pr-2"
            src="https://api.dicebear.com/5.x/avataaars/svg?seed=sarahedo&style=circle"
          />
          Sarah Edo
        </div>
        <div
          class="p-2 flex items-center"
          data-testid="nquestion-answered"
        >
          4
        </div>
        <div
          class="p-2 flex items-center"
          data-testid="nquestion-created"
        >
          2
        </div>
      </div>
      <div
        class="grid grid-flow-row grid-cols-3 gap-4 boder divide-x"
        data-testid="user-row"
      >
        <div
          class="p-2 flex items-center"
        >
          <img
            alt="avatar"
            class="w-10 h-10 rounded-full pr-2"
            src="https://api.dicebear.com/5.x/avataaars/svg?seed=tylermcginnis&style=circle"
          />
          Tyler McGinnis
        </div>
        <div
          class="p-2 flex items-center"
          data-testid="nquestion-answered"
        >
          2
        </div>
        <div
          class="p-2 flex items-center"
          data-testid="nquestion-created"
        >
          2
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
